### ISSUE TITLE
Randomly getting kicked from a multi room while playing

### ISSUE THREAD

**Issue by rednir** [ppy/osu#14365]
**Describe the bug:** 
People seem to get kicked from multiplayer rooms whilst in the middle a playing a match, followed with the notification "Connection to multiplayer server was lost."

Happened about 3-ish times within 40 minutes (and it wasn't just me getting disconnected)

**osu!lazer version:** 
2021.815.0

**Logs:**
[runtime.log](https://github.com/ppy/osu/files/7001692/runtime.log)
[network.log](https://github.com/ppy/osu/files/7001853/network.log)

**Comment by peppy** [comment]
From logs:

```csharp
2021-08-17 16:44:12 [error]: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (An unexpected error occurred invoking 'ChangeUserMods' on the server. InvalidStateException: Incompatible mods were selected: HR,PF,SO,CL,SI,GR,BR,MU)
```

**Comment by frenzibyte** [comment]
Looks to be `OsuModGrow` somehow got in the mix with `OsuModSpinIn`, which are both incompatible. I'll look into this further.

**Comment by frenzibyte** [comment]
The mod incompatibility portion would be a duplicate of #12874

https://user-images.githubusercontent.com/22781491/129829793-4bab7e3b-bef9-4360-8af0-fd8efb58ebc3.mp4

```
[runtime] 2021-08-18 02:57:24 [error]: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (An unexpected error occurred invoking 'ChangeUserMods' on the server. InvalidStateException: Incompatible mods were selected: GR,TC)
[runtime] 2021-08-18 02:57:24 [error]: ---> Microsoft.AspNetCore.SignalR.HubException: An unexpected error occurred invoking 'ChangeUserMods' on the server. InvalidStateException: Incompatible mods were selected: GR,TC
[runtime] 2021-08-18 02:57:24 [error]: at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsyncCore(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
[runtime] 2021-08-18 02:57:24 [error]: at System.Threading.Tasks.ForceAsyncAwaiter`1.GetResult()
[runtime] 2021-08-18 02:57:24 [error]: at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsync(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
[runtime] 2021-08-18 02:57:24 [error]: --- End of inner exception stack trace ---
[runtime] 2021-08-18 02:57:24 [error]: An unobserved error has occurred. (inner)
[runtime] 2021-08-18 02:57:24 [error]: Microsoft.AspNetCore.SignalR.HubException: An unexpected error occurred invoking 'ChangeUserMods' on the server. InvalidStateException: Incompatible mods were selected: GR,TC
[runtime] 2021-08-18 02:57:24 [error]: at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsyncCore(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
[runtime] 2021-08-18 02:57:24 [error]: at System.Threading.Tasks.ForceAsyncAwaiter`1.GetResult()
[runtime] 2021-08-18 02:57:24 [error]: at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsync(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
```

However, I'm not sure what exactly has got you kicked from the multiplayer room as that seems different from the mod incompatibility failure which doesn't affect anything.

Looking at the logs again, it seems you've indeed got booted out while in gameplay:
```
021-08-17 16:46:02 [verbose]: Polling adjusted (listing: 0, selection: 0)
2021-08-17 16:46:02 [verbose]: Screen changed → PlayerLoader
2021-08-17 16:46:02 [verbose]: Polling adjusted (listing: 0, selection: 0)
2021-08-17 16:46:02 [verbose]: Polling adjusted (listing: 0, selection: 0)
2021-08-17 16:46:06 [verbose]: Screen changed → MultiplayerPlayer
2021-08-17 16:46:22 [important]: Connection to multiplayer server was lost.
2021-08-17 16:46:22 [verbose]: Screen changed ← PlayerLoader
2021-08-17 16:46:22 [verbose]: Screen changed ← Multiplayer
2021-08-17 16:46:22 [verbose]: Polling adjusted (listing: 0, selection: 0)
2021-08-17 16:46:22 [verbose]: Polling adjusted (listing: 15000, selection: 15000)
2021-08-17 16:46:23 [verbose]: Focus contention triggered by NotificationOverlay.
```

And the network logs show that the connection was closed without any exceptions thrown, oddly:
```
2021-08-17 16:46:22 [verbose]: OnlineMultiplayerClient disconnected
2021-08-17 16:46:22 [verbose]: OnlineMultiplayerClient connecting...
2021-08-17 16:46:22 [verbose]: Performing request osu.Game.Online.Rooms.SubmitRoomScoreRequest
2021-08-17 16:46:22 [verbose]: Request to https://osu.ppy.sh/api/v2/rooms/97272/playlist/296848/scores/1257860 successfully completed!
2021-08-17 16:46:22 [verbose]: Performing request osu.Game.Online.Rooms.PartRoomRequest
2021-08-17 16:46:22 [verbose]: OnlineMultiplayerClient connected!
2021-08-17 16:46:22 [verbose]: Request to https://osu.ppy.sh/api/v2/rooms/97272/users/17204559 successfully completed!
2021-08-17 16:46:22 [verbose]: Performing request osu.Game.Online.Rooms.GetRoomsRequest
2021-08-17 16:46:23 [verbose]: Request to https://osu.ppy.sh/api/v2/rooms successfully completed!
```

**Comment by rednir** [comment]
I played multi again with one of the players that were having issues, and nobody seemed to be disconnecting while that player did several times. They had `Connection to multiplayer server was lost` in their notifications but *also* `API connection was lost, can't continue with online play`. Unfortunately I don't have their logs, but this just sounds like their internet connection is dropping for a couple seconds. 

So I don't think this as much of an issue as I made it out to be, that's my bad. But I do find it strange that when I disconnected, in my network.log file above there are no timeout errors. Either way you might want to consider closing this.

**Comment by peppy** [comment]
Would definitely help to get more logs of these issues. I think we do need to better handle this situation, but first need to understand what's happening. SignalR should be able to recover from disconnections itself.

**Comment by TheThirdDave** [comment]
A friend and I were experiencing this issue on a semi-regular basis this week. Is there still a need for logs on the issue? We did notice it seemed to be only happening to the host of the multiplayer room from our observations.

I can try to reproduce the issue and attach the logs here.

**Comment by smoogipoo** [comment]
Yes, we do still need logs.

**Comment by Doxxz** [comment]
Regular occurence to me. I've had it happen when in game, when spectating, and on song select.
I suspected it was a Linux only issue, but I asked around and it seems Windows 11 users also get it, not sure about Win10.

Logs:
[network.log](https://github.com/ppy/osu/files/8374680/network.log)
[runtime.log](https://github.com/ppy/osu/files/8374681/runtime.log)

**Comment by peppy** [comment]
For anyone reporting these issue, it would be useful to know your network setup.

- Do you have any proxy servers (is your router doing any kind of ad blocking or otherwise?)
- Do you use any kind of vpn?
- Is there anything between you and the internet?

**Comment by TheThirdDave** [comment]
Sorry for delay.

No proxy or vpn setup. Friend that it happened to is on wifi and I'm hardwired to the router.

Issue occurred mid song while we were playing.

It's happened to both of us pretty evenly. Both are Win11.

These logs are my friends logs that I was playing with to try and recreate the issue. I was able to track down in the logs to the exact time of the issue timestamp: `2022-03-31 00:23:59`

[runtime.log](https://github.com/ppy/osu/files/8385016/runtime.log)
[network.log](https://github.com/ppy/osu/files/8385019/network.log)

Let me know if you need anymore info or would like us to try and do specific things to try and reproduce!

**Comment by peppy** [comment]
@TheThirdDave to confirm, did both you and your friend get disconnected at the same time?

**Comment by TheThirdDave** [comment]
@peppy no just my friend was disconnected.

**Comment by smoogipoo** [comment]
> the logs show that this is a websocket disconnection

There are 3 different logs above, only one of which has the exception there. Hard to say if the others are also websocket disconnection or something else low-level in the TCP stack (OS/hardware).

> Client-side will require some kind of queueing

I would probably not do queueing. I think it is unexpected from a UX point of view to suddenly have 20 actions firing off because the button wasn't clicking.

> how do we guarantee that the client is in the correct state on reconnection

Not having queueing also makes this relatively simple - `JoinRoom` already puts the room in the correct state.

**Comment by peppy** [comment]
> Not having queueing also makes this relatively simple - JoinRoom already puts the room in the correct state.

Yeah, for now that's the case server-side. We'll need to consciously keep that true going forward, if we want to be sure that's the case.

**Comment by peppy** [comment]
As an update, as of the last release the multiplayer endpoint has been rerouted in a way such that disconnections to be about as rare as one would expect from osu!(stable) / bancho.

Going to remove the priority from this issue for now. We may eventually want to revisit this in the future.
